### PR TITLE
rgw: forwarded some requests to master zone

### PR DESCRIPTION
**Pull request by liuchang0812**
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by liuchang0812** [comment]
@cbodley mind taking a look?

File: src/rgw/rgw_rest_s3.cc [hunk 1]
```diff
@@ -2602,6 +2602,8 @@ int RGWSetRequestPayment_ObjStore_S3::get_params()
     return r;
   }
 
+  in_data.append(data);
```
Comment (cbodley) [inline src/rgw/rgw_rest_s3.cc]: can SetRequestPayment use `data` directly, instead of adding a separate `in_data`?

**Comment by cbodley** [comment]
@liuchang0812 can you please add `Fixes: http://tracker.ceph.com/issues/39549` to at least one of the commits?

**Comment by liuchang0812** [comment]
maybe we should abstract `need forward/forward to master zone` logic to reduce that duplicated code.

File: src/rgw/rgw_rest_s3.cc [hunk 2]
```diff
@@ -2595,22 +2595,22 @@ int RGWSetRequestPayment_ObjStore_S3::get_params()
   const auto max_size = s->cct->_conf->rgw_max_put_param_size;
 
   int r = 0;
-  bufferlist data;
-  std::tie(r, data) = rgw_rest_read_all_input(s, max_size, false);
+  std::tie(r, in_data) = rgw_rest_read_all_input(s, max_size, false);
```
Comment (cbodley) [inline src/rgw/rgw_rest_s3.cc]: ah, i didn't realize `data` was a local variable here. looks good :+1:

**Comment by cbodley** [comment]
> maybe we should abstract `need forward/forward to master zone` logic to reduce that duplicated code.

i'd welcome ideas here, especially if they can help us catch this when adding new ops

**Comment by cbodley** [comment]
passed qa in http://pulpito.front.sepia.ceph.com/cbodley-2019-06-19_19:30:04-rgw-wip-cbodley-testing-distro-basic-smithi/. s3test failures tracked in http://tracker.ceph.com/issues/40416, and valgrind warnings in ceph-mon are from https://github.com/ceph/ceph/pull/27834

**Comment by cbodley** [comment]
jenkins test docs

**Comment by cbodley** [comment]
jenkins test docs

**Pull request merged**